### PR TITLE
add net8.0 target

### PR DESCRIPTION
**Pull request by westin-m**
Put CustomKey and a custom converter for SecurityKey in TestUtils, transition aot test to net8.0

### CODE REVIEW

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 1]
```diff
@@ -98,7 +98,7 @@ public bool MapInboundClaims
             }
             set
             {
-                if(!_mapInboundClaims && value && _inboundClaimTypeMap.Count == 0)
+                if (!_mapInboundClaims && value && _inboundClaimTypeMap.Count == 0)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: try not to make these spacing changes, it is just noise and may make integration into dev harder.

Comment (westin-m) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: Must have been something my editor did; I did not manually change any spacing. I'll clean it up.

File: build/commonTest.props [hunk 2]
```diff
@@ -18,11 +18,16 @@
     <RuntimeFrameworkVersion Condition=" '$(TargetFramework)' == 'netcoreapp2.1'">$(DotNetCoreAppRuntimeVersion)</RuntimeFrameworkVersion>
   </PropertyGroup>
 
-  <PropertyGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' OR '$(TargetFramework)' == 'net6.0' ">
+  <PropertyGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' OR '$(TargetFramework)' == 'net6.0' Or '$(TargetFramework)' == 'net8.0'">
```
Comment (jennyf19) [inline build/commonTest.props:21]: >$(TargetFramework)' == 'netcoreapp2.1' [](http://example.com/codeflow?start=30&length=38)

Why is netcoreapp2.1 still used?

File: build/dependencies.props [hunk 3]
```diff
@@ -15,10 +14,19 @@
...
     <SystemSecurityClaimsVersion>4.3.0</SystemSecurityClaimsVersion>
-    <SystemSecurityCryptographyCngVersion>4.5.0</SystemSecurityCryptographyCngVersion>
     <SystemTextEncoding>4.3.0</SystemTextEncoding>
     <SystemTextJson>4.7.2</SystemTextJson>
     <SystemTextEncodingsWeb>4.7.2</SystemTextEncodingsWeb>
     <SystemXmlXmlDocumentVersion>4.3.0</SystemXmlXmlDocumentVersion>
   </PropertyGroup>
+
+  <PropertyGroup Condition="'$(TargetFramework)' == 'net8.0'">
+    <SystemSecurityCryptographyCngVersion>6.0.0-preview.4.21253.7</SystemSecurityCryptographyCngVersion>
```
Comment (jennyf19) [inline build/dependencies.props]: >-preview.4.21253.7 [](http://example.com/codeflow?start=47&length=18)

This there a non-preview which works w/.NET 8? #Resolved

**Comment by jennyf19** [comment]
> #if NET45

If we no longer have net45 support do we need the NET45 ifdefs?

---
Refers to: src/Microsoft.IdentityModel.Tokens/AsymmetricSignatureProvider.cs:190 in d37cb32. [](commit_id = d37cb32be017b62b923461a69dd35b53b7db66a5, deletion_comment = False)

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/Exceptions/SamlSecurityTokenException.cs [hunk 4]
```diff
@@ -17,15 +17,17 @@ public class SamlSecurityTokenException : SecurityTokenException
         /// </summary>
         public SamlSecurityTokenException()
             : base()
-        {}
+        {
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/Exceptions/SamlSecurityTokenException.cs]: I would move this back as they were, they like to have as few lines used as possible. Less noise on the PR as well. #Resolved

**Comment by jennyf19** [comment]
> using Microsoft.IdentityModel.Tokens;

nit: add license info

---
Refers to: test/Microsoft.IdentityModel.TestUtils/CustomSecurityKey.cs:1 in d37cb32. [](commit_id = d37cb32be017b62b923461a69dd35b53b7db66a5, deletion_comment = False)

**Comment by jennyf19** [comment]
> #if NET8_0_OR_GREATER

nit: add license info


---
Refers to: test/Microsoft.IdentityModel.TestUtils/SecurityKeyCustomConverter.cs:1 in d37cb32. [](commit_id = d37cb32be017b62b923461a69dd35b53b7db66a5, deletion_comment = False)

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/Exceptions/SamlSecurityTokenException.cs (hunk 4 above)
Comment (westin-m) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml/Exceptions/SamlSecurityTokenException.cs]: Reverted this as part of the changes I think my editor made that Brent pointed out

File: test/Microsoft.IdentityModel.Tokens.Tests/AsymmetricAdapterTests.cs [hunk 5]
```diff
@@ -24,7 +24,7 @@ public void AsymmetricAdapterUsageTests(AsymmetricAdapterTheoryData theoryData)
 
             try
             {
-#if NET461 || NET462 || NET472 || NETCOREAPP2_1 || NET6_0
+#if NET461 || NET462 || NET472 || NETCOREAPP2_1 || NET6_0 || NET8_0
```
Comment (jennyf19) [inline test/Microsoft.IdentityModel.Tokens.Tests/AsymmetricAdapterTests.cs]: >NETCOREAPP2_1  [](http://example.com/codeflow?start=34&length=14)

I don't think we need netcoreapp2.1 anymore?

**Review by jennyf19 — CHANGES_REQUESTED** [review]
:clock1:

File: test/Microsoft.IdentityModel.TestUtils/SecurityKeyCustomConverter.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
+﻿#if NET8_0_OR_GREATER
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.TestUtils/SecurityKeyCustomConverter.cs]: needs a copyright header

File: build/commonTest.props (hunk 2 above)
Comment (westin-m) [reply build/commonTest.props:21]: I don't remember any conversation about removing it. Should that be a different PR?

File: build/dependencies.props (hunk 3 above)
Comment (westin-m) [reply build/dependencies.props]: Looks like 5.0.0 works. Use that instead?

File: build/commonTest.props (hunk 2 above)
Comment (jennyf19) [reply build/commonTest.props:21]: @brentschmaltz what are your thoughts? I think a separate PR is fine, but let's at least open an issue for tracking

**Comment by westin-m** [comment]
> If we no longer have net45 support do we need the NET45 ifdefs?

No, probably not. That's about 29 more files; should this be its own PR? 

---
In reply to: [1583577448](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2108#issuecomment-1583577448) [](http://example.com/codeflow?ancestors=1583577448)

File: build/dependencies.props (hunk 3 above)
Comment (jennyf19) [reply build/dependencies.props]: They don't have a 6 out of preview? Yes to 5.0.0 then. @brentschmaltz thoughts?

Comment (westin-m) [reply build/dependencies.props]: No, all the 6s are preview right now

**Comment by jennyf19** [comment]
Yes. let's do a separate PR. Please open an issue in wilson for tracking and link to the previous PR for removing the target as well. thanks! sorry to have just noticed that.

---
In reply to: [1583577448](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2108#issuecomment-1583577448) [](http://example.com/codeflow?ancestors=1583577448,1583577448)

File: build/dependencies.props (hunk 3 above)
Comment (jennyf19) [reply build/dependencies.props]: Let's ship with the latest stable GA version.

File: build/commonTest.props (hunk 2 above)
Comment (jmprieur) [reply build/commonTest.props:21]: Could it be to test the special case of aspnetcore2.1 on .NET FW (still maintained)

Comment (jennyf19) [reply build/commonTest.props:21]: nevermind. we discussed offline.

File: test/Microsoft.IdentityModel.Tokens.Tests/AsymmetricAdapterTests.cs (hunk 5 above)
Comment (jennyf19) [reply test/Microsoft.IdentityModel.Tokens.Tests/AsymmetricAdapterTests.cs]: discussed offline

**Review by jennyf19 — APPROVED** [review]

File: build/targets.props [hunk 7]
```diff
@@ -1,6 +1,6 @@
 <Project>
   <PropertyGroup>
-    <SrcTargets>net461;net462;net472;netstandard2.0;net6.0</SrcTargets>
+    <SrcTargets>net461;net462;net472;netstandard2.0;net6.0;net8.0</SrcTargets>
```
Comment (jmprieur) [inline build/targets.props:3]: Maybe we should only enable the net8.0 target framework if people opt-in by setting an environment variable in their console or machine `USE_DOTNET_8_PREVIEW`

```suggestion
    <SrcTargets Condition="'$(USE_DOTNET_8_PREVIEW)' == 'true' ">net461;net462;net472;netstandard2.0;net6.0;net8.0</SrcTargets>
```

File: build/targetsTest.props [hunk 8]
```diff
@@ -1,6 +1,6 @@
 <Project>
   <PropertyGroup>
-    <TestTargets>net461;net462;net472;netcoreapp2.1;net6.0</TestTargets>
+    <TestTargets>net461;net462;net472;netcoreapp2.1;net6.0;net8.0</TestTargets>
```
Comment (jmprieur) [inline build/targetsTest.props:3]: Same comment.
What do you think, @brentschmaltz, @jennyf19, @westin-m, @keegan-caruso 
This avoid forcing the community to install a preview of .NET 8, if they don't want to?

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 9]
```diff
@@ -1165,7 +1165,7 @@ internal IEnumerable<SecurityKey> GetContentEncryptionKeys(JsonWebToken jwtToken
             {
                 try
                 {
-#if NET472 || NET6_0
+#if NET472 || NET6_0 || NET8_0
```
Comment (jmprieur) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: Can't we use `NET6_0_OR_GREATER`.
See the table in https://learn.microsoft.com/dotnet/csharp/language-reference/preprocessor-directives#conditional-compilation

**Review by jmprieur — COMMENTED** [review]
LGTM, but I've not tried locally
I've added  a few suggestions

File: build/targetsTest.props (hunk 8 above)
Comment (jennyf19) [reply build/targetsTest.props:3]: Yes, I like that. I think we had talked about doing that before, no?

Comment (westin-m) [reply build/targetsTest.props:3]: Yes, we did talk about it and I like the idea. Brent pushed back, saying that it's more work on the developer's part.

File: build/targets.props (hunk 7 above)
Comment (westin-m) [reply build/targets.props:3]: This is in its own issue now https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2111

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 9 above)
Comment (westin-m) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: Good suggestion. Done.

**Pull request merged**